src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.hpp

```diff
@@ -0,0 +1,279 @@
+/*!
+ * \brief Computes the scalar wave upwind multipenalty boundary
+ * correction.
+ *
+ * This implements the upwind multipenalty boundary correction term
+ * \f$D_\alpha\f$. The general form is given by:
+ *
+ * \f{align*}{
+ *   \label{eq:pnpm upwind boundary term characteristics}
+ *   D_\beta =
+ *   T_{\beta\hat{\beta}}^{\mathrm{ext}}
+ *   \Lambda^{\mathrm{ext},-}_{\hat{\beta}\hat{\alpha}}
+ *   v^{\mathrm{ext}}_{\hat{\alpha}}
+ *   -T_{\beta\hat{\beta}}^{\mathrm{int}}
+ *   \Lambda^{\mathrm{int},-}_{\hat{\beta}\hat{\alpha}}
+ *   v^{\mathrm{int}}_{\hat{\alpha}},
+ * \f}
+ *
+ * We denote the evolved fields by \f$u_{\alpha}\f$, the characteristic fields
+ * by \f$v_{\hat{\alpha}}\f$, and implicitly sum over reapeated indices.
+ * \f$T_{\alpha\hat{\alpha}}\f$ transforms characteristic fields to evolved
+ * fields, while \f$\Lambda_{\hat{\alpha}\hat{\beta}}^-\f$ is a diagonal matrix
+ * with only the negative characteristic speeds. The int and ext superscripts
+ * denote quantities on the internal and external side of the mortar. Note that
+ * Eq. (6.3) of \cite Teukolsky2015ega is not exactly what's implemented since
+ * that boundary term does not consistently treat both sides of the interface on
+ * the same footing.
+ *
+ * For the scalar wave system the correction is:
+ *
+ * \f{align}{
+ *   D_{\Psi} &= \tilde{\lambda}_{v^{\Psi}}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},\Psi}
+ *                - \tilde{\lambda}_{v^{\Psi}}^{\mathrm{int}}
+ *                v^{\mathrm{int},\Psi}, \\
+ *   D_{\Pi} &= \frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{ext}}
+ *             v^{\mathrm{ext},+} +
+ *             \tilde{\lambda}_{v^-}^{\mathrm{ext}}
+ *             v^{\mathrm{ext},-}\right)
+ *             + \tilde{\lambda}_{v^\Psi}^\mathrm{ext}\gamma_2
+ *             v^{\mathrm{ext},\Psi}
+ *             \notag \\
+ *           &-\frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{int}}
+ *             v^{\mathrm{int},+} +
+ *             \tilde{\lambda}_{v^-}^{\mathrm{int}}
+ *             v^{\mathrm{int},-}\right)
+ *             - \tilde{\lambda}_{v^\Psi}^\mathrm{int}\gamma_2
+ *             v^{\mathrm{int},\Psi} , \\
+ *   D_{\Phi_{i}}
+ *              &= \frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},+}
+ *                - \tilde{\lambda}_{v^-}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},-}\right)n_i^{\mathrm{ext}}
+ *                + \tilde{\lambda}_{v^0}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},0}_{i}
+ *                \notag \\
+ *              &-
+ *                \frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{int}}
+ *                v^{\mathrm{int},+}
+ *                - \tilde{\lambda}_{v^-}^{\mathrm{int}}
+ *                v^{\mathrm{int},-}\right)n_i^{\mathrm{int}}
+ *                - \tilde{\lambda}_{v^0}^{\mathrm{int}}
+ *                v^{\mathrm{int},0}_{i},
+ * \f}
+ *
+ * with characteristic fields
+ *
+ * \f{align}{
```

Comment:
might be nice to note the storage convention for the char speeds -- it makes sense, but it's not immediately obvious.

Reply:
As in, which component maps to which speed?

Reply:
Yeah, exactly.